walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -13,11 +13,118 @@
 import java.util.Collection;
+import java.util.Collections;
+import java.util.Set;
+import java.util.HashSet;
+import java.util.List;
+import java.util.ArrayList;
+import java.util.Arrays;
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    List<TimeRange> ret = new ArrayList<TimeRange>();
+
+    Set<String> mandatoryAttendees =  new HashSet<String>();
+    mandatoryAttendees.addAll(request.getAttendees()); 
+
+    Set<String> optionalAttendees = new HashSet<String>();
+    optionalAttendees.addAll(request.getOptionalAttendees());
+
+    Set<String> allAttendees = new HashSet<String>();
+    allAttendees.addAll(mandatoryAttendees);
+    allAttendees.addAll(optionalAttendees);
+
+    long duration = request.getDuration();
+
+    // no options for too long of a request
+    if (duration > 24*60){
+      return ret;
+    }
+
+    // options for no attendees
+    if (allAttendees.size() == 0) {
+      ret.add(TimeRange.WHOLE_DAY);
+      return ret;
+    }
+
+    // find free times
+    List<TimeRange> freeTimes = new ArrayList<TimeRange>(findFreeTimes(allAttendees, events, request));
+    if (freeTimes.isEmpty()) {
+      // ignore optional attendees if there are no free times across both mandatory and optional attendees
+      freeTimes = new ArrayList<TimeRange>(findFreeTimes(mandatoryAttendees, events, request));
+    } else {
+      return freeTimes;
+    }
+
+    // adhere to additional test requirement
+    if (mandatoryAttendees.isEmpty()) {
+      return new ArrayList<TimeRange>();
+    }
+    return freeTimes;
+  }
+
+  private List<TimeRange> mergeOverlaps(List<TimeRange> blocked) {
+    // find if blocked times have overlap and consolidate them into sorted arraylist
+    Collections.sort(blocked, TimeRange.ORDER_BY_START);
+    List<TimeRange> merged = new ArrayList<TimeRange>();
+    merged.add(blocked.get(0));
+    int idx = 0;
+    for (int i = 0; i < blocked.size(); i++) {
+      if (blocked.get(i).overlaps(merged.get(idx))) {
+        merged.set(idx, TimeRange.fromStartEnd(
+          Math.min(merged.get(idx).start(), blocked.get(i).start()),
+          Math.max(merged.get(idx).end(), blocked.get(i).end()) - 1, 
+          true));
+      } else {
+        idx += 1;
+        merged.add(blocked.get(i));
+      }
+    }
+    return merged;
+  }
+
+  private List<TimeRange> findFreeTimes(Set<String> attendees, Collection<Event> events, MeetingRequest request) {
+    // loop through blocked times, add free time ranges to ret
+    List<TimeRange> freeTimes = new ArrayList<TimeRange>();
+    List<TimeRange> blockedTimes = new ArrayList<TimeRange>();
+    long duration = request.getDuration();
+
+    for (Event event : events) {
+      Set<String> eventAttendees = event.getAttendees();
+      Set<String> overlappingAttendees = new HashSet<>(eventAttendees);
```

Comment:
I think creating a set for every event is a great approach. sets are designed to have extremely fast lookup times (calling contains()) is very efficient, but the creation of sets is not as efficient (new HashSet<>(eventAttendees) will create a new hash set, then add each attendee by iterating through eventAttendees)

to fix this, we can try to avoid creating sets, and only try looking up values. something like this will work better:
for(Event event : events) { 
  if(event.getAttendees().stream().anyMatch(attendees::contains)) {
    blockedTimes.add(event.getWhen());
  }
}

Reply:
Ah interesting! I didn't know streams existed in Java